tests/acceptance/test_update.py

```diff
@@ -139,24 +184,133 @@ def test_network_based_image_update(self, successful_image_update_mender):
+            self.test_network_based_image_update(successful_image_update_mender, bitbake_variables)
+
+            (active, passive) = determine_active_passive_part(bitbake_variables)
+            assert(active == bitbake_variables["MENDER_ROOTFS_PART_B"])
+
+        # Make a note of the checksums of each environment. We use this later to
+        # determine which one changed.
+        old_checksums = Helpers.get_env_checksums(bitbake_variables)
+
+        orig_env = run("fw_printenv")
+
+        # Make a dummy/broken update
+        subprocess.call("dd if=/dev/zero of=image.dat bs=1M count=0 seek=8", shell=True)
+        subprocess.call("artifacts write rootfs-image -t %s -n test-update -u image.dat -o image.mender" % image_type, shell=True)
+        put("image.mender", remote_path="/var/tmp/image.mender")
+        run("mender -rootfs /var/tmp/image.mender")
+
+        new_checksums = Helpers.get_env_checksums(bitbake_variables)
+
+        # Exactly one checksum should be different.
+        assert(old_checksums[0] == new_checksums[0] or old_checksums[1] == new_checksums[1])
+        assert(old_checksums[0] != new_checksums[0] or old_checksums[1] != new_checksums[1])
+
+        if old_checksums[0] != new_checksums[0]:
+            changed = 0
+        elif old_checksums[1] != new_checksums[1]:
+            changed = 1
+        else:
+            assert False, "Should not happen"
+
+        if corrupt_old_env:
+            # We should corrupt the old environment. Unlikely to happen in
+            # practice, but should be recoverable as well.
+            to_corrupt = 1 - changed
+        else:
+            # We should corrupt the new environment.
+            to_corrupt = changed
+
+        offsets = Helpers.get_env_offsets(bitbake_variables)
+
+        # Now manually corrupt the environment.
+        # A few bytes should do it!
+        run("dd if=/dev/zero of=%s bs=1 count=64 seek=%d oflag=sync"
+            % (bitbake_variables["MENDER_STORAGE_DEVICE"], offsets[to_corrupt]))
+
+        # Check atomicity of Mender environment update: The contents of the
+        # environment before the update should be identical to the environment
+        # we get if we update, and then corrupt the new environment. If it's not
+        # identical, it's an indication that there were intermediary steps. This
+        # is important to avoid so that the environment is not in a half updated
+        # state.
+        new_env = run("fw_printenv")
+        if not corrupt_old_env:
+            # Only relevant when corrupting the new environment.
+            assert orig_env == new_env
+
+        reboot()
+
+        # We should have recovered.
+        run_after_connect("true")
+
+        # And we should be back at the second rootfs partition.
+        (active, passive) = determine_active_passive_part(bitbake_variables)
+        assert(active == bitbake_variables["MENDER_ROOTFS_PART_B"])
+
+        # Cleanup.
```

Comment:
There is a lot of assertions here, so possibly we will exit without cleaning this. If we don't care and everything will be wiped out after the test this is OK. If this is not the case, maybe we should close this inside `try` and then do `except AssertionError:` here.